File: src/machines/javascript/scripts/lib-story.js
// begin: Story Module -------------------------------------------------------

assets.load_story = function(path) {
	const text = assets.read_text_file(path);
	const story = {
		 _text: text,
		 _callbacks: [],
		 path: path,
		 moments: JSON.parse(text),
		 cursor: 0,
		 current_moment: null
	}; 

	// find the index of a moment by its id, returns -1 on fail.
	story.find_index = function(moment_id) {
		 if (moment_id == null) return -1;
		 for (var i=0; i<story.moments.length; i++) {
			  if (story.moments[i].id && story.moments[i].id == moment_id) {
					return i;
			  }
		 }
		 return -1;
	};

	// move the cursor to a specific moment (by id).
	story.goto_moment = function(moment_id) {
		 var index = moment_id == null ? 0 : story.find_index(moment_id);
		 if (index >= 0) {
			  story.goto_index(index);
		 }
	}

	// goto a specific moment index.
	story.goto_index = function(index) {
		 story.cursor = index;
		 story.current_moment = story.moments[index];
		 // trigger prompt event.
		 story.trigger('moment', story.current_moment);
	}

	// move the cursor to the next moment.
	story.next = function() {
		 var next = story.cursor + 1;
		 if (next < story.moments.length) {
			  story.goto_index(next);
		 }
	}

	// completely reset the story.
	story.reset = function() {
		 story.moments = JSON.parse(text);
		 story.trigger('reset', null);
		 story.goto_index(0);
	}

	// trigger a story event.
	story.trigger = function(event, data) {
		 if (story._callbacks[event] !== undefined) {
			  const cb = story._callbacks[event];
			  for (var i=0; i<cb.length; i++) {
					cb[i](data);
			  }
		 }
	}

	// handle various story events (moment, reset)
	story.on = function(event, cb) {
		 if (story._callbacks[event] === undefined) {
			  story._callbacks[event] = [];
		 }
		 story._callbacks[event].push(cb);
	}

	story.start = function() {
		 story.goto_moment(null);
	}
	
	return story;
}

// end: Story Module ---------------------------------------------------------

File: src/machines/lua/scripts/lib-story.lua
-- start: Story Module -------------------------------------------------------

assets.load_story = function(path)
	text = assets.read_text_file(path)
	story = {
		 _text = text,
		 _callbacks = {},
		 path = path,
		 moments = json.parse(text),
		 cursor = 0,
		 current_moment = nil
	}

	-- find the index of a moment by its id, returns 0 on fail.
	story.find_index = function(moment_id)
		 if (moment_id == null) then return 0 end
		 for i=1,story.moments.length do
			  if story.moments[i].id and story.moments[i].id == moment_id then
					return i
			  end
		 end
		 return 0
	end

	-- move the cursor to a specific moment (by id).
	story.goto_moment = function(moment_id)
		 index = moment_id == nil and 1 or story.find_index(moment_id)
		 if index >= 0 then
			  story.goto_index(index)
		 end
	end

	-- goto a specific moment index.
	story.goto_index = function(index)
		 story.cursor = index
		 story.current_moment = story.moments[index]
		 -- trigger prompt event.
		 story.trigger('moment', story.current_moment)
	end

	-- move the cursor to the next moment.
	story.next = function()
		 next = story.cursor + 1
		 if story.moments[next] ~= nil then
			  story.goto_index(next)
		 end
	end

	-- completely reset the story.
	story.reset = function()
		 story.moments = json.parse(text)
		 story.trigger('reset', null)
		 story.goto_index(1)
	end

	-- trigger a story event.
	story.trigger = function(event, data)
		 if story._callbacks[event] ~= nil then
			  cb = story._callbacks[event]
			  for i,v in ipairs(cb) do
					cb[i](data)
			  end
		 end
	end

	-- handle various story events (moment, reset)
	story.on = function(event, cb)
		 if not story._callbacks[event] ~= nil then
			  story._callbacks[event] = {}
		 end
		 table.insert(story._callbacks[event], cb)
	end

	story.start = function()
		 story.goto_moment(null)
	end
	
	return story
end

-- end: Scene Module ---------------------------------------------------------